[Docs] Add user level documentation for Expressions (#11990)
KratosMultiphysics/Kratos: docs/pages/Kratos/Expressions/Utilities/Collapse.md

---
title: Collapse
keywords: 
tags: [collapse, lazy expression, expressions]
sidebar: kratos_expressions
summary: 
---

## Introduction

> ```Collapse``` method is used to collapse the lazy expression tree. **This is useful when the lazy expression tree is larger which occupies or exceeds the RAM
> limitations.**.

Reviewer: ```suggestion
The ```Collapse``` method forces an evaluation of the input expression into a flat array. **This is useful when the input expression represents a complex/deep/chained operation with many operands that occupy considerable memory. In such a scenario, `Collapse` allows the user to replace the original expression hierarchy with a single array (freeing up RAM), at the expense of immediate evaluation.**.
```

Author: done :)

> Using this method on an expression which will always result in an expression having the same shape, having the same evaluated values. But the resulting expression will be not having the tree structure since it is collapsed.

Reviewer: ```suggestion
The resulting expression retains the original's shape.
```

Author: done :)

## Use case

> The max depth of an expression can be checked to identify the max depth of the lazy expression as explained in the following code snippet. It also explains how to collapse the tree hierarchy of the lazy expressions.

Reviewer: ```suggestion
The following example shows how to compute the maximum depth of the expression hierarchy, and collapse it down to a single array.
```

Author: done :)

```python
import KratosMultiphysics as Kratos
model = Kratos.Model()
model_part = model.CreateModelPart("test")

node_1 = model_part.CreateNewNode(1, 0, 0, 0)
node_2 = model_part.CreateNewNode(2, 1, 0, 0)
node_3 = model_part.CreateNewNode(3, 1, 1, 0)

# setting VELOCITY of each node
node_1.SetValue(Kratos.VELOCITY, Kratos.Array3([-1,-2,-3]))
node_2.SetValue(Kratos.VELOCITY, Kratos.Array3([-4,-5,-6]))
node_3.SetValue(Kratos.VELOCITY, Kratos.Array3([-7,-8,-9]))

# create the nodal expression
nodal_expression = Kratos.Expression.NodalExpression(model_part)

# read the VELOCITY from non-historical nodal container
Kratos.Expression.VariableExpressionIO.Read(nodal_expression, Kratos.VELOCITY, False)

nodal_expression *= 2
nodal_expression += 1

print(nodal_expression.GetMaxDepth())

> collapsed_exp = Kratos.Expression.Utils.Collapse(nodal_expression)
> print(collapsed_exp.GetMaxDepth())

Reviewer: ```suggestion
collapsed_exp = Kratos.Expression.Utils.Collapse(nodal_expression)

# Destroy the original expression to free up memory
del nodal_expression

print(collapsed_exp.GetMaxDepth())
```

Author: done :)

```
Expected output:
```bash

> 3
> 1
>  |  /           |                  
>  ' /   __| _` | __|  _ \   __|    
>  . \  |   (   | |   (   |\__ \  
> _|\_\_|  \__,_|\__|\___/ ____/
>            Multi-Physics 9.4."3"-docs/expression_documentation-6de5f1a499-Release-x86_64
>            Compiled for GNU/Linux and Python3.11 with GCC-13.2
> Compiled with threading and MPI support.
> Maximum number of threads: 30.
> Running without MPI.

Reviewer: ```suggestion
 |  /           |                  
 ' /   __| _` | __|  _ \   __|    
 . \  |   (   | |   (   |\__ \  
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 9.4."3"-docs/expression_documentation-6de5f1a499-Release-x86_64
           Compiled for GNU/Linux and Python3.11 with GCC-13.2
Compiled with threading and MPI support.
Maximum number of threads: 30.
Running without MPI.
3
1
```

Author: done :)
